### ISSUE TITLE
required=False and default is not empty no Longer Allows KeyError

### ISSUE THREAD

**Issue by cancan101** [encode/django-rest-framework#2785]
Currently I have a field:

``` python
is_foo = serializers.BooleanField(source='app_metadata.is_foo', required=False)
```

which works fine to parse a dict looking like:
`{ }`, `{'app_metadata':{}}` or `{'app_metadata': {'is_foo': True}}`.

If I then change the Field to:

``` python
is_foo = serializers.BooleanField(source='app_metadata.is_foo', required=False, default=False)
```

I can no longer handle: `{ }` or `{'app_metadata':{}}` which seems counter intuitive / wrong.

The [offending code](https://github.com/tomchristie/django-rest-framework/blob/56ec7b152e0d88933973ae4da84b563c446042a6/rest_framework/fields.py#L301):

```
    def get_attribute(self, instance):
...
        try:
            return get_attribute(instance, self.source_attrs)
        except (KeyError, AttributeError) as exc:
            if not self.required and self.default is empty:
                raise SkipField()
...
            raise type(exc)(msg)
```

I would think that a missing field should be OK regardless of whether or not there is a default. The only difference should be that the default is inserted when missing.

I get the same issue with the Field:

``` python
is_foo = serializers.BooleanField(source='app_metadata.is_foo', default=False)
```

**Comment by kived** [comment]
*BUMP*

I'm dealing with this as well. I use OrientDB, so it's possible for a given field to not exist in a record - particularly since newly added properties don't appear on existing records. I have many fields with `(required=False, read_only=True)`, which all work fine. But changing that to `(required=False, read_only=True, default=False)` (or any `default`, this example specifically is a `BooleanField`) causes everything to break. I would expect DRF to return the default value if an actual value doesn't exist, since that's generally what defaults are for.

My current workaround:
```
def __init__(self, instance=None, data=empty, **kwargs):
    super(GCSerializer, self).__init__(instance=instance, data=data, **kwargs)
    if instance:
        for name, field in self.fields.iteritems():
            if field.default is not empty:
                instance.setdefault(name, field.default)
```

My record instances are just `dict`s.

**Comment by TBeijen** [comment]
Ran into this issue as well, as we're using serializers to normalize blobs of json of which attributes might be missing.

I'm especially interested in what the rationale was of requiring a key or attribute, even if default has been set, as that might be a scenario I'm overlooking and unaware of.

**Comment by xordoquy** [comment]
What's the meaning of setting `required=False` and `default` ?
I'm not sure why you'd want to set a default value if the field isn't required.

**Comment by xordoquy** [comment]
> I can no longer handle: { } or {'app_metadata':{}} which seems counter intuitive / wrong.

What's the current outcome here ? What's expected ?

**Comment by TBeijen** [comment]
Expected: When serializing use default if missing from instance. This behaviour replaces omitting of the field.
```
from rest_framework import serializers

class ExampleSerializer(serializers.Serializer):
	is_foo = serializers.BooleanField(source='app_metadata.is_foo', required=False, default=False)
	has_default = serializers.CharField(default='abc')

ExampleSerializer({}).data
```

Trying to set both a default and required=True, raises as an error, as it did before this change:

```
class WrongSerializer(serializers.Serializer):
	is_bar = serializers.BooleanField(required=True, default='abc')

Traceback (most recent call last):
<TRUNCATED>
AssertionError: May not set both `required` and `default`
```

I suppose the question boils down to: Is there a case where one wants a default to be used when deserializing raw data to instance (if not partial updating) but does not want the default to be returned when serializing an instance that has not data for the field. If so, then this needs reconsidering.

**Comment by TBeijen** [comment]
Adding some observations that might help getting this out of 'needs review' (be it via approval or identifying further steps to be taken).

The use case this change tries to solve is serializing non-model objects that might/might not have certain attributes. Examples would be serializing document structures stored in NoSQL dbs or JSON fields, where the document's structure has changed over time.

The change doesn't affect converting representation to internal value.
The change doesn't impact serializing Django models, as they have a well-defined set of attributes, so serializing a Django model will never invoke the default value mechanism.

**Comment by tomchristie** [comment]
My previous responses have been that default is intended to be about data validation behavior, not about filling in missing attributes during serialization. However given that there's more and more cases for non-model serialization I think we could be willing to review this, given a decent pull request to drive that change.

**Comment by TBeijen** [comment]
@tomchristie Do you consider my PR https://github.com/tomchristie/django-rest-framework/pull/4692 a good fix? If not, I'm happy to fix any shortcomings.

**Comment by carltongibson** [comment]
This was fixed by #4692, v3.6.0
